src/priority-queue.ts

```diff
@@ -0,0 +1,257 @@
+import {ArmorCollection} from './collection';
+import {ArmorCollectionSelector} from './selector';
+import ArmorPriorityQueueComparator from './priority-queue-comparator';
+import ArmorPriorityQueueNodeChildren from './priority-queue-children';
+import ArmorPriorityQueueOptions from './priority-queue-options';
+
+export default class ArmorPriorityQueue<T> implements ArmorCollection<T> {
+	public elements: T[];
+	public readonly comparator: ArmorPriorityQueueComparator<T>;
+
+	constructor(comparator: ArmorPriorityQueueComparator<T>, options?: ArmorPriorityQueueOptions<T>) {
+		if (typeof comparator !== 'function') {
+			throw new Error('Must have a comparator function for priority queue to operate properly');
+		}
+
+		this.elements = [];
+		this.comparator = comparator;
+
+		if (options) {
+			this.parseOptions(options);
+		}
+	}
+
+	public parseOptions(options: ArmorPriorityQueueOptions<T>): void {
+		if (options.elements !== undefined) {
+			if (!Array.isArray(options.elements)) {
```

Comment:
The above check for options.elements !== undefined can actually be eliminated due to using the isArray check. isArray will return false when options.elements is undefined or if its defined as anything else other than an array meaning both cases are handled by one check.  `options.elements` being `undefined` will not throw when called or used, but `options.elements` **will** throw when called if `options` is undefined.

Reply:
Also note, you can change the isArray check to be an early drop if this function only parses `elements`.  If multiple independent properties are validated, early dropping out of one will cancel validation of all following checks which isn't necessarily what you want. In this case we would create validation functions for each property checked if any work goes into validation and call each one inside `parseOptions`. So, `parseOptionsElements`, `parseOptionsId` (made that one up), etc.